binlog.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "binlog.h"
+
+#include <assert.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+struct _packed_ binlog_msg_header {
+    uint16_t msg_size;
+    uint8_t msg_type;
+};
+
+static bool _binlog_timeout_cb(void *data)
+{
+    BinLog *binlog = static_cast<BinLog *>(data);
+    return binlog->logging_start_timeout();
+}
+
+bool BinLog::logging_start_timeout()
+{
+    mavlink_message_t msg;
+
+    mavlink_msg_remote_log_block_status_pack(_system_id, MAV_COMP_ID_ALL, &msg, _target_system_id, MAV_COMP_ID_ALL, MAV_REMOTE_LOG_DATA_BLOCK_START, 1);
+    _send_msg(&msg, _target_system_id);
+
+    return true;
+}
+
+bool BinLog::start()
+{
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    char *filename;
+    int r;
+
+    if (_file != -1) {
+        log_warning("BinLog already started");
+        return false;
+    }
+
+    r = asprintf(&filename, "%s/%i-%02i-%02i_%02i-%02i-%02i.bin", _logs_dir,
+                 timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday,
+                 timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec);
+    if (r < 1) {
+        log_error_errno(errno, "Error formatting BinLog file name: (%m)");
+        return false;
+    }
+
+    _file = open(filename, O_WRONLY | O_CLOEXEC | O_CREAT | O_NONBLOCK | O_TRUNC,
+                 S_IRUSR | S_IROTH | S_IRGRP);
+    if (_file < 0) {
+        log_error_errno(errno, "Unable to open BinLog file(%s): (%m)", filename);
+        goto open_error;
+    }
+
+    // TODO check why in a timeout
+    // TODO ulog also check for liveness - should we do the same here?
+    _timeout = _mainloop->add_timeout(MSEC_PER_SEC, _binlog_timeout_cb, this);
```

Comment:
You never stops this timeout... you should do it as soon you got an ack from MAV_REMOTE_LOG_DATA_BLOCK_START.

Also you should check for liveness as user can reboot the flight stack anytime.